lambda-buffers-compiler/src/LambdaBuffers/Compiler/KindCheck/Inference.hs

```diff
@@ -0,0 +1,413 @@
+{-# LANGUAGE TemplateHaskell #-}
+{-# OPTIONS_GHC -Wno-missing-import-lists #-}
+{-# OPTIONS_GHC -Wno-missing-kind-signatures #-}
+{-# OPTIONS_GHC -Wno-missing-local-signatures #-}
+{-# OPTIONS_GHC -Wno-missing-signatures #-}
+{-# OPTIONS_GHC -Wno-monomorphism-restriction #-}
+{-# OPTIONS_GHC -Wno-redundant-constraints #-}
+{-# OPTIONS_GHC -Wno-unused-top-binds #-}
+
+module LambdaBuffers.Compiler.KindCheck.Inference (
+  getType,
+  Kind (..),
+  Context (..),
+  Atom,
+  Type (..),
+  infer,
+  DeriveM,
+  DeriveEff,
+  InferErr (..),
+  context,
+  addContext,
+) where
+
+import Data.Bifunctor (Bifunctor (second))
+
+import Control.Monad.Freer
+import Control.Monad.Freer.Error
+import Control.Monad.Freer.Reader
+import Control.Monad.Freer.State
+import Control.Monad.Freer.Writer
+
+import Control.Lens (Getter, makeLenses, to, (&), (.~), (^.))
+
+import Prettyprinter
+
+type Atom = String
+infixr 8 :->:
+
+data Kind
+  = Type
+  | Kind :->: Kind
+  | KVar String
```

Comment:
Imo, this introduces PolyKinds, ie. `a :: forall k. k`, which we explicitly don't support right? So we could remove it?

Reply:
I'll address this comment as a to do in the next version.